src/attributes.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {documentInfoForDoc} from './document-info.js';
+import {viewerForDoc} from '../src/viewer';
+import {getLengthNumeral} from '../src/layout';
+import {getMode} from './mode';
+
+/**
+ * Produces the attributes for the ad template.
+ * @param {!Window} parentWindow
+ * @param {!Element} element
+ * @return {!Object} Contains
+ *       precedence over the data- attributes.
+ *     - data-* attributes of the <amp-ad> tag with the "data-" removed.
+ *     - A _context object for internal use.
+ */
+export function getContextMetadata(
```

Comment:
I can remove a fair amount (which I have done, but it's not in this pr). If you'd like that all to be in this PR, I can do that, but getting this attributes.js file in for the next release is important for us fixing an issue with a4a so I didn't want to slow down if there were any issues integrating this file into 3p-frame.js